### PR TITLE
Fix sort_kanban N+1, use us.total_points, add outer try (2.3.2)

### PR DESCRIPTION
**Pull request by hemati**
## Summary

Production was hitting `sort_kanban_by_rice_tool` and getting the harness-generic `Error occurred during tool execution` with three different request_ids — no JSON payload, so no diagnostic. K8s pod logs showed: liveness probe on `/mcp/health` timed out (`context deadline exceeded`), kubelet killed the container, the in-flight tool response died with the worker.

Root cause: serial N+1. With ~40 user stories + epics, the tool was making `1 + N + 1 + 2M + 1` HTTP calls (`list_user_stories` + per-story `get_attributes` + `list_epics` + per-epic `get_attributes` + per-epic `list_user_stories` + `list_points`). At ~100ms each that's ~5–10s wall time, well past the FastMCP worker's liveness budget while the worker was holding the request.

## Four fixes

1. **Use `us.total_points` directly.** Taiga's `/userstories?project=X` list endpoint already inlines the per-role-summed effort. Eliminates one HTTP call (`list_points()`) and removes the points-resolution code path that bug-cycled twice in 2.3.0 / 2.3.1.

2. **Group by epic in-memory from inline `us.epics`.** No more per-epic `epic.list_user_stories()` — completion% is computed from the already-loaded stories list. Eliminates `O(epics)` extra HTTP calls.

3. **Parallelize the unavoidable `get_attributes()` calls** with `ThreadPoolExecutor(max_workers=10)`. python-taiga's sync calls are thread-safe for read-side GETs (each thread reuses the underlying `requests.Session`). Wall time on the per-US RICE-attribute fetch drops from `O(N)` to `O(1)`.

4. **Outer `try/except Exception`** wrapping the whole function body so any uncaught failure returns a JSON 500 with `trace_tail` (last 5 traceback lines) instead of bubbling up as the generic "Error occurred during tool execution". This is the diagnostic safety net that was missing — the prod issue was undiagnosable without pod logs.

## What changed

- `langchain_taiga/tools/taiga_tools.py:sort_kanban_by_rice_tool`: rewritten data-fetch section. Sort/group/bulk-update logic at the bottom unchanged but now indented inside the outer try.
- `tests/unit_tests/test_sort_kanban_by_rice_tool.py`: existing tests updated for the new `total_points`-based effort path, plus 2 new tests:
  - `test_effort_zero_when_no_points_assigned` — edge case for unestimated stories (`total_points=None` → effort=0, RICE=0).
  - `test_outer_try_returns_json_on_unexpected_failure` — regression guard so the harness-generic-error issue doesn't recur.
- `pyproject.toml`: 2.3.1 → 2.3.2.

Existing test `test_works_when_developer_role_is_not_id_19` renamed to `test_works_regardless_of_role_id` since the new code doesn't iterate role-IDs at all (was meaningful when 2.3.0 removed the hardcoded "19", less so now that we read `total_points` directly).

## Validation

- `make test`: **173 passed, 1 skipped** (was 171 pre-PR, +2 new tests).
- Probed live `wahed` to confirm the data shape: `/userstories?project=X` does NOT include `attributes_values` inline (so the per-US `get_attributes` is unavoidable, hence the parallelization). It DOES include `total_points`, `epics`, `points`, `status`, `swimlane`, `due_date`, `is_closed` — every other field the tool needs.
- Decided NOT to use raw HTTP: python-taiga's `Resource.get_attributes()` is the cleanest path for the one remaining N+1 step, and `ThreadPoolExecutor` over the python-taiga API is enough.

## Notable risks

- **Behaviour preservation**: same effort semantics as 2.3.1 (sum across computable role assignments), just delegated to Taiga's server-side `total_points` field instead of computed locally. No user-visible change in sort order.
- **Threading**: python-taiga's `Requester` reuses a `requests.Session`, which is thread-safe for read GETs. Write paths are not parallelized, so no risk of concurrent-write races.
- **`max_workers=10`**: deliberate cap to keep load on the Taiga API server bounded. Even on projects with hundreds of stories, total wall time stays under K8s liveness probe budget.

## Test plan

- [x] `make test` (173 passed).
- [x] Outer-try regression locked by `test_outer_try_returns_json_on_unexpected_failure`.
- [ ] Smoke test in prod after merge + deploy: `sort_kanban_by_rice_tool("wahed")` should return JSON cleanly within seconds, regardless of how many user stories `wahed` has. If it still 500s, the JSON now carries the actual cause via `trace_tail`.

### CODE REVIEW

File: langchain_taiga/tools/taiga_tools.py [hunk 1]
```diff
@@ -2375,90 +2381,117 @@ def sort_kanban_by_rice_tool(
...
-                    closed = sum(1 for us in related_us if getattr(us, "is_closed", False))
-                    epic_completions[epic.id] = closed / total
-                else:
-                    epic_completions[epic.id] = 0.0
+                return us.ref, us.get_attributes().get("attributes_values", {}) or {}
             except Exception:
-                epic_completions[epic.id] = 0.0
-    except Exception:
-        pass  # If we can't get epics, continue without multiplicators
+                return us.ref, {}
```
[the remark this answers is not in the document]
Comment (hemati) [reply langchain_taiga/tools/taiga_tools.py]: Fixed in commit `6216d69`. You're right — silent default to `1×1×1=1` shoves a real high-priority story to the bottom of its column with no signal to the caller. Two changes:

1. **Total-failure guard:** if every per-story fetch failed, the tool aborts with 500 + `attribute_fetch_errors` listing the cause refs. RICE scores would be uniformly garbage and reordering from defaults is actively harmful.

2. **Partial-failure surfacing:** for the (more common) case where a few fetches fail transiently, the tool now records each failed `{"ref": <us-ref>, "error": "<msg>"}` in `attribute_fetch_errors` of the success payload (key is `None` when every fetch succeeded). The sort still runs but the caller can see exactly which stories' RICE was unreliable and decide to retry or reorder by hand.

Decided against full abort on first failure (the pre-2.3.2 behavior) because that turns one transient timeout out of 40 stories into "whole sort fails", which was its own UX issue.

Locked by `test_partial_attr_fetch_failure_is_surfaced` and `test_total_attr_fetch_failure_returns_500`.

File: langchain_taiga/tools/taiga_tools.py [hunk 2]
```diff
@@ -2375,90 +2381,117 @@ def sort_kanban_by_rice_tool(
...
+
+        with ThreadPoolExecutor(max_workers=10) as executor:
+            us_attr_values = dict(executor.map(_fetch_us_attrs, stories))
```
[the remark this answers is not in the document]
Comment (hemati) [reply langchain_taiga/tools/taiga_tools.py]: Fixed in commit `6216d69` — went with both your suggestions in combination:

- **(b)** Success payload carries `attribute_fetch_errors` field listing `{"ref": <us-ref>, "error": "<msg>"}` for any failed fetches (None when all succeeded). Caller sees partial failures without aborting the rest of the sort.
- **(a)** If literally every fetch failed, the tool returns 500 with the error list instead of sorting from useless defaults — that case is special because RICE numbers would be uniformly garbage and reordering would be actively harmful.

Two new tests lock both paths: `test_partial_attr_fetch_failure_is_surfaced` (1-of-N fails) and `test_total_attr_fetch_failure_returns_500` (N-of-N fails). Full suite: 175 passed.
